### PR TITLE
chore(config): config/profiles refactor for simplicity

### PR DESCRIPTION
**Pull request by ctrombley**

### CODE REVIEW

File: internal/config/config.go [hunk 1]
```diff
@@ -63,379 +134,504 @@ func (c *Value) IsDefault() bool {
...
 
-	output.Printf("%s set to %s\n", text.Bold.Sprint(key), text.FgCyan.Sprint(value))
+	log.Debugf("could not get profile value %s, using default value", key)
+	return ""
+}
 
-	return nil
+func GetActiveProfileName() string {
+	defaultProfile := defaultProfileName()
+	if ProfileOverride != "" {
```
Comment (ctrombley) [inline internal/config/config.go:247]: Invert if

File: internal/config/config.go [hunk 2]
```diff
@@ -63,379 +134,504 @@ func (c *Value) IsDefault() bool {
...
-	err = cfgViper.WriteConfigAs(path)
-	if err != nil {
+	credsFilePath := path.Join(ConfigDir, credsFilename)
+	if err := p.WriteConfigAs(credsFilePath); err != nil {
 		return err
 	}
 
+	if defaultProfileName() == profileName {
+		log.Infof("unsetting %s as default profile.", profileName)
+		if err := SaveDefaultProfileName(""); err != nil {
```
Comment (ctrombley) [inline internal/config/config.go:346]: Delete the file rather than setting empty string

File: internal/config/config.go [hunk 3]
```diff
@@ -63,379 +134,504 @@ func (c *Value) IsDefault() bool {
...
 
-		return nil
-	})
-	if err != nil {
-		log.Error(err)
+func FatalIfAccountIDNotPresent() int {
+	v := GetActiveProfileAccountID()
+	if v == 0 {
+		f := findProfileField(AccountID)
+		log.Fatalf("%s is required, set it in your default profile or use the %s environment variable", AccountID, f.EnvOverride)
```
Comment (ctrombley) [inline internal/config/config.go:373]: return err here

File: internal/config/config.go [hunk 4]
```diff
@@ -63,379 +134,504 @@ func (c *Value) IsDefault() bool {
...
 
-func (c *Config) set(key string, value interface{}) error {
-	cfgViper, err := readConfig(c.configDir)
-	if err != nil {
-		return err
+func FatalIfActiveProfileFieldIntNotPresent(key ProfileFieldKey) int {
+	v := GetActiveProfileValueInt(key)
+	if v == 0 {
+		f := findProfileField(key)
+		log.Fatalf("%s is required, set it in your default profile or use the %s environment variable", key, f.EnvOverride)
```
Comment (ctrombley) [inline internal/config/config.go:383]: return err here

File: internal/config/config.go [hunk 5]
```diff
@@ -63,379 +134,504 @@ func (c *Value) IsDefault() bool {
...
+}
 
-	allScopes, err := unmarshalAllScopes(cfgViper)
-	if err != nil {
-		return err
+func FatalIfActiveProfileFieldStringNotPresent(key ProfileFieldKey) string {
+	v := GetActiveProfileValueString(key)
+	if v == "" {
+		f := findProfileField(key)
+		log.Fatalf("%s is required, set it in your default profile or use the %s environment variable", key, f.EnvOverride)
```
Comment (ctrombley) [inline internal/config/config.go:393]: return err here

File: internal/config/config.go [hunk 6]
```diff
@@ -63,379 +134,504 @@ func (c *Value) IsDefault() bool {
...
 	}
 
-	if err := mergo.Merge(c, defaultConfig); err != nil {
+	return nil
+}
+
+func saveDefaultProfileName(profileName string) error {
+	defaultProfileFilePath := filepath.Join(ConfigDir, defaultProfileFilename)
+
+	if err := ioutil.WriteFile(defaultProfileFilePath, []byte("\""+profileName+"\""), 0644); err != nil {
```
Comment (ctrombley) [inline internal/config/config.go:521]: prefer 0640

File: internal/config/config.go [hunk 7]
```diff
@@ -63,379 +134,504 @@ func (c *Value) IsDefault() bool {
...
 
+func getDefaultConfigDirectory() (string, error) {
+	home, err := homedir.Dir()
 	if err != nil {
-		return err
+		return "", err
 	}
 
-	return nil
+	return fmt.Sprintf("%s/.newrelic", home), nil
```
Comment (ctrombley) [inline internal/config/config.go:542]: move to const

File: internal/config/config.go [hunk 8]
```diff
@@ -63,379 +134,504 @@ func (c *Value) IsDefault() bool {
...
 
 	return nil
 }
 
-func unmarshalAllScopes(cfgViper *viper.Viper) (*map[string]Config, error) {
-	cfgMap := map[string]Config{}
-	err := cfgViper.Unmarshal(&cfgMap)
-	if err != nil {
-		return nil, fmt.Errorf("failed to unmarshal config with error: %v", err)
+func isValidCredentialKey(key ProfileFieldKey) bool {
```
Comment (ctrombley) [inline internal/config/config.go:573]: rename to isValidProfileKey

File: internal/config/config.go [hunk 9]
```diff
@@ -63,379 +134,504 @@ func (c *Value) IsDefault() bool {
...
-		createErr := c.createFile(path, cfgViper)
-		if createErr != nil {
-			return createErr
-		}
-	} else {
-		log.Debugf("writing config file at %s", path)
-		err = cfgViper.WriteConfigAs(path)
-		if err != nil {
-			log.Error(err)
+func ProfileExists(profile string) bool {
```
Comment (ctrombley) [inline internal/config/config.go:434]: Add godoc to all exported methods

File: internal/config/config_test.go [hunk 10]
```diff
@@ -0,0 +1,325 @@
+// +build integration
```
Comment (ctrombley) [inline internal/config/config_test.go:1]: Rename this file to config_integration_test.go

File: internal/edge/command_trace_observer.go [hunk 11]
```diff
@@ -32,6 +32,10 @@ var cmdTraceObserver = &cobra.Command{
 	provides visualization for the most actionable data so you can investigate and
 	solve issues faster.`,
 	Example: "newrelic edge trace-observer list --accountId <accountID>",
+	PreRun: func(cmd *cobra.Command, args []string) {
```
Comment (ctrombley) [inline internal/edge/command_trace_observer.go:35]: Change to PersistentPreRun.  Validate with command metadata tests (same with others)

File: internal/entities/command_tags.go [hunk 12]
```diff
@@ -83,15 +78,13 @@ The delete-values command deletes the specified tag:value pairs on a given entit
...
-		client.WithClient(func(nrClient *newrelic.NewRelic) {
-			tagValues, err := assembleTagValuesInput(entityValues)
-			utils.LogIfFatal(err)
+		tagValues, err := assembleTagValuesInput(entityValues)
+		utils.LogIfFatal(err)
 
-			_, err = nrClient.Entities.TaggingDeleteTagValuesFromEntity(entities.EntityGUID(entityGUID), tagValues)
-			utils.LogIfFatal(err)
+		_, err = client.Client.Entities.TaggingDeleteTagValuesFromEntity(entities.EntityGUID(entityGUID), tagValues)
+		utils.LogIfFatal(err)
```
Comment (ctrombley) [inline internal/entities/command_tags.go:85]: Get rid of LogIfFatal usage

File: internal/entities/command_tags.go [hunk 13]
```diff
@@ -104,15 +97,13 @@ The create command adds tag:value pairs to the given entity.
 `,
 	Example: "newrelic entity tags create --guid <entityGUID> --tag tag1:value1",
 	Run: func(cmd *cobra.Command, args []string) {
-		client.WithClient(func(nrClient *newrelic.NewRelic) {
-			tags, err := assembleTagsInput(entityTags)
-			utils.LogIfFatal(err)
+		tags, err := assembleTagsInput(entityTags)
+		utils.LogIfFatal(err)
```
Comment (ctrombley) [inline internal/entities/command_tags.go:101]: Get rid of LogIfFatal usage

File: internal/entities/command_tags.go [hunk 14]
```diff
@@ -126,15 +117,13 @@ provided for the given entity.
 `,
 	Example: "newrelic entity tags replace --guid <entityGUID> --tag tag1:value1",
 	Run: func(cmd *cobra.Command, args []string) {
-		client.WithClient(func(nrClient *newrelic.NewRelic) {
-			tags, err := assembleTagsInput(entityTags)
-			utils.LogIfFatal(err)
+		tags, err := assembleTagsInput(entityTags)
+		utils.LogIfFatal(err)
```
Comment (ctrombley) [inline internal/entities/command_tags.go:121]: Get rid of LogIfFatal usage

File: internal/entities/command_tags.go [hunk 15]
```diff
@@ -126,15 +117,13 @@ provided for the given entity.
...
 
-			_, err = nrClient.Entities.TaggingReplaceTagsOnEntity(entities.EntityGUID(entityGUID), tags)
-			utils.LogIfFatal(err)
+		_, err = client.Client.Entities.TaggingReplaceTagsOnEntity(entities.EntityGUID(entityGUID), tags)
+		utils.LogIfFatal(err)
```
Comment (ctrombley) [inline internal/entities/command_tags.go:124]: Get rid of LogIfFatal usage

File: internal/nerdstorage/command.go [hunk 16]
```diff
@@ -18,4 +20,8 @@ var (
 var Command = &cobra.Command{
 	Use:   "nerdstorage",
 	Short: "Read, write, and delete NerdStorage documents and collections.",
+	PreRun: func(cmd *cobra.Command, args []string) {
```
Comment (ctrombley) [inline internal/nerdstorage/command.go:23]: Change to PersistentPreRun

File: internal/nrql/command.go [hunk 17]
```diff
@@ -2,10 +2,15 @@ package nrql
...
 	"github.com/spf13/cobra"
+
+	"github.com/newrelic/newrelic-cli/internal/config"
 )
 
 // Command represents the nerdgraph command.
 var Command = &cobra.Command{
 	Use:   "nrql",
 	Short: "Commands for interacting with the New Relic Database",
+	PreRun: func(cmd *cobra.Command, args []string) {
```
Comment (ctrombley) [inline internal/nrql/command.go:13]: Change to PersistentPreRun

File: internal/profiles/command.go [hunk 18]
```diff
@@ -37,31 +47,46 @@ for posting custom events with the ` + "`newrelic events`" + `command.
...
-				APIKey:            apiKey,
-				InsightsInsertKey: insightsInsertKey,
-				AccountID:         accountID,
-				LicenseKey:        licenseKey,
+		if config.ProfileExists(profileName) {
+			log.Fatalf("profile already exists: %s", profileName)
+		}
+
+		if err := config.SaveValueToProfile(profileName, config.APIKey, apiKey); err != nil {
+			if e := config.RemoveProfile(profileName); e != nil {
```
Comment (ctrombley) [inline internal/profiles/command.go:55]: Create update & save commands to cut down on boiler plate

File: internal/profiles/command.go [hunk 19]
```diff
@@ -94,13 +116,56 @@ The list command prints out the available profiles' credentials.
...
+
+		if len(profileNames) == 0 {
+			log.Info("no profiles found")
+			return
+		}
+
+		// Print them out
+		for _, n := range profileNames {
+			var accountIDStr string
+			accountIDVal := config.GetProfileValueInt(n, config.AccountID)
```
Comment (ctrombley) [inline internal/profiles/command.go:130]: refactor to getAccountID

**Review by zlesnr — DISMISSED** [review]

**Review by zlesnr — APPROVED** [review]

**Review by Julien4218 — CHANGES_REQUESTED** [review]
This work is currently been re-done

**Comment by ctrombley** [comment]
Duplicate of #948

**Pull request closed without merging**